### DISCUSSION TITLE
Flask application + uwsgi: SSL SYSCALL EOF detected error

### DISCUSSION THREAD

**Discussion by danner26** [sqlalchemy/sqlalchemy#5956] [Usage Questions] [answered]
**Describe the bug**
I am really sorry if this is not the right place for this, but I am trying to cover all of my bases, throw out everything I have and see if I get a bite anywhere. Any help you may have to offer is greatly appreciated! 

Hello, we are plagued by the never ending SSL SYSCALL error. We are running out postgres database on an RDS using a db.t2.large. Resource wise it has plenty of space and resources, and the connections are set to an amount that is no where near being hit. We have been plagued with this issue for a long time, but I just recently was assigned to the issue. This tends to happen when we the following uwsgi ini:
```
[uwsgi]
binary-path = /usr/sbin/uwsgi
virtualenv = /opt/webapps/iamp/venv
strict = false
module = wsgi
plugins = python36
buffer-size = 16384
die-on-term = true
lazy-apps = false #make this true to fix syscall issue
single-interpreter = true

# NETWORK CONFIG
socket = :8080
chmod-socket = 660

# WORKER CONFIG
master = true
processes = 4 #reduce this to 1 if syscall issue persists
enable-threads = true
#vacuum = true
threads = 8
minspare = 4
```

When we do not use lazy-apps it appears that it causes this issue. Sometimes it wont happen right away, but once you start writing to the database we start getting flooded with these. It breaks our flask application as well on the front end. A full dump of the error page can be seen here: https://pastebin.com/s5CgBrGs and https://pastebin.com/ySeVivHe

When we turn lazy apps on the ssl syscall issue disappears. Any assistance figuring this out would be greatly appreciated. 

**Versions.**
 - OS: RHEL 8.3
 - Python: 3.6
 - SQLAlchemy: 1.3.23
 - Flask-SQLAlchemy version: 2.4.4
 - Database: Postgres 10.13
 - DBAPI:

**Comment by jvanasco** [comment, marked as answer]
`lazy-apps` causes uwsgi to load the application post-fork, instead of pre-fork.

the exception you are seeing is from psycogp2, and usually tied to:

* database connection/disconnect issue
* database is full or tables locked
* database is out of memory

because lazy-apps post-fork affects this, it makes me think you have some code in your application that was connecting to the database before the fork.  

my suggestion is to use the uwsgi postfork decorator (https://uwsgi-docs.readthedocs.io/en/latest/PythonDecorators.html#uwsgidecorators.postfork) and to write a snippet of code that will call `engine.dispose`  (https://docs.sqlalchemy.org/en/13/core/connections.html#engine-disposal)

whoops, closed by accident. sorry about that.

based on your description, this sounds like a very likely issue is that you are connecting to the database before the fork, and the post-fork exceptions on "write" activity is happening against database connections that do not exist, are being used by another process, or are in a closed state.

**Reply by zzzeek** [reply]
just FYI this looks like a fastAPI question, "lazy-apps" has nothing to do with SQLAlchemy.

set the pool type to NullPool and that will remove anything to do with connection pooling from the equation.

**Reply by danner26** [reply]
Thanks for the tips, I have tried the NullPool before and I keep seeing "SSL error: decryption failed or bad record mac" in turn. Any idea what that is @zzzeek - I have tried troubleshooting that issue as well without any success

**Reply by danner26** [reply]
I am assuming that what @jvanasco said, it is because of the code in the app. Something related to this?
https://stackoverflow.com/questions/22752521/uwsgi-flask-sqlalchemy-and-postgres-ssl-error-decryption-failed-or-bad-reco

**Reply by jvanasco** [reply]
@danner26 yes that’s the same issue. Down the thread, someone gives the “correct” answer to use ‘dispose’, as the lazy methods just keep your app from running efficiently.

you can try turning on SqlAlchemy logging and do a startup/few requests/shutdown then examining the logs.  That should help you pinpoint when/where there is a prefork connection.

90% of SqlAlchemy problems I’ve seen like this are because of prefork connections. Usually an app is either needlessly connecting to the database too early, or is loading some application setup defaults.  Disposal solves this, as does refactoring code so those situations don’t happen.   For the pyramid framework, I maintain a postfork hook library to solve this (the crypto module needs to be reseeded too, iirc)

If you are loading app config data at startup, you can also use a different db connection string and engine, then trash them after loading. 

Because Mike’s suggestion presented a new problem, I would personally audit how and where sessions are created and accessed.  It seems like your app has an anti pattern in it somewhere, and there is some connection/data-flow happening outside of the expected request lifecycle.  If the app is older, it might also be using SqlAlchemy wrong, and not properly closing connections or something similar.  I have seen a handful of related problems caused by a framework/plug-in not implementing some of this correctly, but by the time they fix their problem and docs... users have already deployed the problematic code/patterns.

**Reply by zzzeek** [reply]
is there any concurrent access to connections or cursors going on ?   this seems like a question for psycopg2 people to be looking into.   i have no information here.

**Reply by danner26** [reply]
Thank you, I appreciate the information. You gave me a lot to look into and I will have to figure out what is going on at a deeper level, then I will contact psycopg2 if needed. Thanks again

**Reply by danner26** [reply]
Also as a heads up the NullPool fixes the issue completely, there was a modification I made during testing that I forgot to remove before I tried the nullpool. I would like to get the pooling working but for now this will work.
